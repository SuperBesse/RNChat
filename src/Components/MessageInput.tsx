import React, { useEffect, useRef } from 'react';
import { TextInput, StyleSheet, ActivityIndicator, View } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { sendMessage } from '@/Redux/Actions/MessageActions';
import { AppState } from '@/Redux/Reducers';
type Props = {};

const MessageInput = ({}: Props) => {
  const [text, setText] = React.useState<string>('');
  const textInputRef = useRef<TextInput>(null);
  //listen text changing but rerender only if text has changed
  const onChangeText = (value: string) => {
    if (value !== text) {
      setText(value);
    }
  };

  const { t } = useTranslation();
  const dispatch = useDispatch();

  //get status onf message to display loader and reset text on success
  const isSending = useSelector((state: AppState) => {
    const isLoading = state.messageState.isSending;
    if (isLoading) {
      onChangeText('');
    }
    return isLoading;
  });

  useEffect(() => {
    if (!isSending && textInputRef.current) {
      textInputRef.current.focus();
    }
  }, [isSending]);

  const sendNewMessage = () => {
    dispatch(sendMessage(text));
  };

  return (
    <View style={styles.container}>
      <TextInput
        style={styles.input}
        ref={textInputRef}
        onChangeText={onChangeText}
        value={text}
        placeholder={t('placeholder.message')}
        onEndEditing={sendNewMessage}
        editable={!isSending}
        enablesReturnKeyAutomatically
        autoCorrect={false}
        autoCapitalize={'none'}
      />
      {isSending && (
        <ActivityIndicator
          size="small"
          color="#grey"
          testID="activity-sending"
        />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  input: {
    alignSelf: 'center',
    borderColor: 'gray',
    width: '80%',
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginVertical: 8,
    marginRight: 8,
  },
});

export default MessageInput;
