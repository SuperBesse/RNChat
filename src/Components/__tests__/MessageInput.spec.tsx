import React from 'react';
import { act, render, screen } from '@testing-library/react-native';
import MessageInput from '../MessageInput';
import * as redux from 'react-redux';
import { TextInput } from 'react-native';
import { put, takeEvery } from 'redux-saga/effects';
import addMessageSaga, { addMessage } from '@/Redux/Sagas/AddMessage';
//import { expectSaga } from 'redux-saga-test-plan';
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

jest.mock('../../Translations');
const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
afterAll(() => useDispatchSpy.mockClear());

describe('MessageInput', () => {
  let generator = null;
  beforeAll(() => {
    generator = addMessageSaga();
  });
  it('should render correctly', () => {
    const tree = render(<MessageInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should action send message on onEndEditing event', async () => {
    render(<MessageInput />).toJSON();

    const addMessagePayload = {
      payload: '',
      type: 'SEND_MESSAGE_START',
    };
    act(() => {
      screen.UNSAFE_getByType(TextInput).props.onEndEditing();
    });
    expect(mockDispatch).toHaveBeenCalledWith(addMessagePayload);
    useDispatchSpy.mockClear();
  });

  it('should trigger message saga when message action is send', async () => {
    put({ type: 'SEND_MESSAGE_START' });
    const actual = generator.next();
    expect(actual.value).toEqual(takeEvery('SEND_MESSAGE_START', addMessage));

    // return expectSaga(addMessageSaga)
    //   .put({
    //     type: 'SEND_MESSAGE_SUCCESS',
    //     message: 'message test',
    //   })
    //   .dispatch({
    //     type: 'SEND_MESSAGE_SUCCESS',
    //     payload: { message: 'message test' },
    //   })
    //   .run();
  });
});
