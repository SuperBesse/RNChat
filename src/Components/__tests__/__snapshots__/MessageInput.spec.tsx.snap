// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInput should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <TextInput
    editable={true}
    enablesReturnKeyAutomatically={true}
    onChangeText={[Function]}
    onEndEditing={[Function]}
    placeholder="placeholder.message"
    style={
      Object {
        "alignSelf": "center",
        "borderColor": "gray",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginRight": 8,
        "marginVertical": 8,
        "padding": 10,
        "width": "80%",
      }
    }
    value=""
  />
</View>
`;
